README
App-Waf

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::Waf

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-Waf

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-Waf

    CPAN Ratings
        http://cpanratings.perl.org/d/App-Waf

    Search CPAN
        http://search.cpan.org/dist/App-Waf/


LICENSE AND COPYRIGHT

Copyright (C) 2016 ORANGE

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for App-Waf

0.08    2018-02-01/16
* bug fix.

0.07    2017-07-14/16
*  fix iptables can't find bug
*  add scane for SYN attack and ban for IPs

0.06    2016-12-13/18
*  add Parameter fault tolerance.

0.05    2016-11-18/18
* pod doc fix.

0.04    2016-11-18/18
* bug fixes.

0.03    2016-11-17/19
* add iptables banip practice example. 
* add nginx banip practice example.

0.02    2016-11-16/19
* add summery output for sort of type.
* practice examples an example sub.

0.01    2016-11-14/19
*First version, released on an unsuspecting world.



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'App::Waf',
    AUTHOR           => q{ORANGE <bollwarm@ijz.me>},
    VERSION_FROM     => 'lib/App/Waf.pm',
    ABSTRACT_FROM    => 'lib/App/Waf.pm',
    LICENSE          => 'Perl',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
        'File::ReadBackwards' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
    },
   
       META_MERGE => {
          'meta-spec' => { version => 2 },
           resources   => {
            repository => {
                type => 'git',
                url  => 'https://github.com/bollwarm/app-waf.git',
                web  => 'http://ijz.me',
            },
        },
    },


    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'App-Waf-*' },
);


example/banip.pl
#!/usr/bin/perl

use App::Waf;

# 设置日志文件和需要解析的文件大小，一般是web日志，$threshold为封禁的阈值
# 可以根据实际情况调节大小

my $filename  = "/web/logs/access.log";
my $numlines  = 10000;
my $threshold = 200;

=pod
## 结合nginx 和 iptables 进行实时banip的实例（example/banip.pl）

## 加入crontab 每5分钟执行一次。

=code<`echo "*/5 * * * * perl $dir/banip.pl >> bianip.logs 2>&1 " >> /var/spool/cron/root`>
## 以下设置nginx格式，包括nginx封禁的文件和重启nginx
## 需要的pid格式。

=cut

my $nginx_home  = "/usr/local/nginx";
my $ngixBanfile = $nginx_home . '/conf/conf.d/blockip.conf';
my $ngixPidfile = $nginx_home . '/logs/nginx.pid';

my $line = tail( $filename, $numlines );


( $log, $zcount, $zip, $zrequrl, $zstatus, $siteurl ) = initCount($line);

for ( sort { $zip->{$b} <=> $zip->{$a} } keys %{$zip} ) {

    print "$_ : $zip->{$_} \n" if $zip->{$_} > $threshold;

    nginxBan($_,$ngixBanfile,$ngixPidfile) if $zip->{$_} >$threshold;
    #iptabBan( $_, $ngixBanfile, $ngixPidfile ) if $zip->{$_} > $threshold;

}
sub nginxBan {

    my $btime = localtime( time() );
    my ( $ip, $conf, $pid ) = @_;
    my $bid = 0;
    open my $nFD, "<", $conf or die("Can not open the file!$!\n");
    while (<$nFD>) {
        print "DEBUG ::nginxBan :: $conf IN $_" if $DEBUG;
        $bid = 1 if /$ip/;
    }
    close $nFD;

    open my $nFD, ">>", $conf or die("Can not open 1 the file!$!\n");

    unless ($bid) {
        print "$btime,banip $ip\n";
        print $nFD "deny $ip\;\n";
        $pid=`cat $pid`;
        chomp $pid;
         `/usr/bin/kill -HUP $pid`;
    }

    close $nFD;

    #print  "/bin/kill -HUP $nginx_home/logs/nginx.pid \n";
}

sub iptabBan {

# must be root user;
# 必须root用户才可以操作iptables，当然也必须有iptables服务跑动着

    my $IP = shift;
    
    my $ips = `iptables-save`;
    my @ipsline=split /\n/sm,$ips;
    my $dist=0;
    for(@ipsline){
        
      $dist=1 if (/$IP/ and /INPUT/ and /DROP/);

     }
    unless ($dist) {
        `iptables -I INPUT -s $IP -j DROP`;
        my $btime = localtime( time() );
        print "$btime :band $IP \n";
    } else {
   
     print "band alread!\n";

    }

}


example/fwban.pl
#!/usr/bin/perl

use App::Waf;

# 设置日志文件和需要解析的文件大小，一般是web日志，$threshold为封禁的阈值
# 可以根据实际情况调节大小

my $filename  = "/web/logs/access.log";
my $numlines  = 10000;
my $threshold = 200;

=pod
## 结合nginx 和 iptables 进行实时banip的实例（example/banip.pl）

## 加入crontab 每5分钟执行一次。

=code<`echo "*/5 * * * * perl $dir/banip.pl >> bianip.logs 2>&1 " >> /var/spool/cron/root`>
## 以下设置nginx格式，包括nginx封禁的文件和重启nginx
## 需要的pid格式。

=cut

my $nginx_home  = "/usr/local/nginx";
my $ngixBanfile = $nginx_home . '/conf/conf.d/blockip.conf';
my $ngixPidfile = $nginx_home . '/logs/nginx.pid';

my $line = tail( $filename, $numlines );

( $log, $zcount, $zip, $zrequrl, $zstatus, $siteurl ) = initCount($line);

for ( sort { $zip->{$b} <=> $zip->{$a} } keys %{$zip} ) {

    print "$_ : $zip->{$_} \n" if $zip->{$_} > $threshold;

    iptabBan( $_, $ngixBanfile, $ngixPidfile ) if $zip->{$_} > $threshold;

}


sub iptabBan {

# must be root user;
# 必须root用户才可以操作iptables，当然也必须有iptables服务跑动着

    my $IP = shift;

    my $ips     = `/sbin/iptables-save`;
    my @ipsline = split /\n/sm, $ips;
    my $dist    = 0;
    for (@ipsline) {

        $dist = 1 if ( /$IP/ and /INPUT/ and /DROP/ );

    }
    unless ($dist) {
        `/sbin/iptables -I INPUT -s $IP -j DROP`;
        my $btime = localtime( time() );
        print "$btime :band $IP \n";
    }
    else {

        print "band alread!\n";

    }

}


example/synban.pl
#!/usr/bin/perl

use App::Waf;

# 设置日志文件和需要解析的文件大小，一般是web日志，$threshold为封禁的阈值
# 可以根据实际情况调节大小

=pod
## 结合nginx 和 iptables 进行实时banip的实例（example/banip.pl）

## 加入crontab 每5分钟执行一次。

=code<`echo "*/5 * * * * perl $dir/banip.pl >> bianip.logs 2>&1 " >> /var/spool/cron/root`>

=cut

my $cmd=q(/usr/sbin/ss -a|grep SYN-RECV|perl -lane 'print $F[-1]'|perl -pe 's/:.*$//');
my @sync_ps_count=`$cmd`;
my %ipcount;
my $threshold=8;
my $btime=localtime time;
for(@sync_ps_count){

chomp;
$ipcount{$_}++;
}

for(sort {$ipcount{$b}<=>$ipcount{$a}} keys %ipcount) {

print "$btime $_ SYN攻击次数: $ipcount{$_} \n";

# Count are more than shreshold, Then Ban it through iptables;

if ($ipcount{$_} > $threshold) {

  print  "$btime Ban The IP ：$_ \n";

  iptabBan($_);
}

}

=pod
sub iptabBan {

# must be root user;
# 必须root用户才可以操作iptables，当然也必须有iptables服务跑动着

    my $IP = shift;

    my $ips     = `/sbin/iptables-save`;
    my @ipsline = split /\n/sm, $ips;
    my $dist    = 0;
    for (@ipsline) {

        $dist = 1 if ( /$IP/ and /INPUT/ and /DROP/ );

    }
    unless ($dist) {
        `/sbin/iptables -I INPUT -s $IP -j DROP`;
        my $btime = localtime( time() );
        print "$btime :band $IP \n";
    }
    else {

        print "band alread!\n";

    }

}
=cut


example/ngixban.pl
#!/usr/bin/perl

use App::Waf;

# 设置日志文件和需要解析的文件大小，一般是web日志，$threshold为封禁的阈值
# 可以根据实际情况调节大小

my $filename  = "/web/logs/access.log";
my $numlines  = 10000;
my $threshold = 200;

=pod
## 结合nginx 和 iptables 进行实时banip的实例（example/banip.pl）

## 加入crontab 每5分钟执行一次。

=code<`echo "*/5 * * * * perl $dir/banip.pl >> bianip.logs 2>&1 " >> /var/spool/cron/root`>
## 以下设置nginx格式，包括nginx封禁的文件和重启nginx
## 需要的pid格式。

=cut

my $nginx_home  = "/usr/local/nginx";
my $ngixBanfile = $nginx_home . '/conf/conf.d/blockip.conf';
my $ngixPidfile = $nginx_home . '/logs/nginx.pid';

my $line = tail( $filename, $numlines );

( $log, $zcount, $zip, $zrequrl, $zstatus, $siteurl ) = initCount($line);

for ( sort { $zip->{$b} <=> $zip->{$a} } keys %{$zip} ) {

    print "$_ : $zip->{$_} \n" if $zip->{$_} > $threshold;

    nginxBan( $_, $ngixBanfile, $ngixPidfile ) if $zip->{$_} > $threshold;


}

=pod
sub nginxBan {

    my $btime = localtime( time() );
    my ( $ip, $conf, $pid ) = @_;
    my $bid = 0;
    open my $nFD, "<", $conf or die("Can not open the file!$!\n");
    while (<$nFD>) {
        print "DEBUG ::nginxBan :: $conf IN $_" if $DEBUG;
        $bid = 1 if /$ip/;
    }
    close $nFD;

    open my $nFD, ">>", $conf or die("Can not open 1 the file!$!\n");

    unless ($bid) {
        print "$btime,banip $ip\n";
        print $nFD "deny $ip\;\n";
        $pid = `cat $pid`;
        chomp $pid;
        `/usr/bin/kill -HUP $pid`;
    }

    close $nFD;

}
=cut


example/summary.pl
use App::Waf;

my $filename = "example.acess";

# change with you practice log pash,such as :
#$filename =  "/var/logs/httpd/access.log";
my $numlines = shift;

my $line = tail( $filename, $numlines );

my ( $log, $zcount, $zip, $zrequrl, $zstatus, $siteurl ) = initCount($line);

print "==============Attack Summary ==================\n";
print "\nThe total attack count: $zcount \n";
print "\nThe count from source IP:  \n\n";
print "$_\=> $zip->{$_} \n" for ( sort keys %{$zip} );
print "The count From request Url:  \n\n";
print "$_\=> $zrequrl->{$_} \n" for ( sort keys %{$zrequrl} );
print "\n\nThe count From Http Status:  \n\n";
print "$_\=> $zstatus->{$_} \n" for ( sort keys %{$zstatus} );
print "\n\nThe count From Site Url:  \n\n";
print "$_\=> $siteurl->{$_} \n" for ( sort keys %{$siteurl} );

#print $log;


lib/App/Waf.pm
package App::Waf;

use 5.006;
use strict;
use warnings;
require Exporter;

=encoding utf8
=head1 NAME

App::Waf - A sample  Web Application Firewall,
analysis the web logs for illegal attempt in real time。
summary the source IP and other tpyes infomations ,using
this infomations for ban whith iptables. 

通过解析web访问日志，实时统计非法访问，结合防火期等进行
主动式防御。 

=head1 VERSION

Version 0.08

=cut

our $VERSION = '0.08';

our @ISA    = qw(Exporter);
our @EXPORT = qw(tail initCount iptabBan nginxBan);

=head1 SYNOPSIS
=head2 实例

  use App::Waf;
  my $filename = "example.acess";#日志文件
  my $numlines  = 50000; #要处理的行数,从后读。
  my $line=tail($filename,$$numlines);
   ($log,$zcount,$zip,$zrequrl,$zstatus,$siteurl)=initCount($line);
  print "==============Attack Summary ==================\n";
  print "\nThe total attack count: $zcount \n";
  print "\nThe count from source IP:  \n\n";
  print "$_\=> $zip->{$_} \n" for(sort  keys %{$zip});
  print "The count From request Url:  \n\n";
  print "$_\=> $zrequrl->{$_} \n" for(sort keys %{$zrequrl});
  print "\n\nThe count From Http Status:  \n\n";
  print "$_\=> $zstatus->{$_} \n" for(sort keys %{$zstatus});
  print "\n\nThe count From Site Url:  \n\n";
  print "$_\=> $siteurl->{$_} \n" for(sort keys %{$siteurl});
  
=head2 结合nginx 和 iptables 进行实时banip的实例（example/banip.pl）

加入crontab 每5分钟执行一次。

   echo "*/5 * * * * perl $dir/banip.pl >> bianip.logs 2>&1 " >> /var/spool/cron/root

=head1 SUBROUTINES/METHODS

=head2 tail()

IN: $logfile,$count;

OUT: return the the latest $count lines of the $logfile.  

=head2 initCount()

IN: the content of need to cheack and count.

OUT: all types count result.

=cut

use File::ReadBackwards;

my $DEBUG = 0;

my @validurl = (
    'rfd.php\?include_file',
    '\.\./',
    'select.+(from|limit)',
    '(?:(union(.*?)select))',
    'having|rongjitest',
    'sleep\((\s*)(\d*)(\s*)\)',
    'benchmark\((.*)\,(.*)\)',
    'base64_decode\(',
    '(?:from\W+information_schema\W)',
    '(?:(?:current_)user|database|schema|connection_id)\s*\(',
    '(?:etc\/\W*passwd)',
    'into(\s+)+(?:dump|out)file\s*',
    'group\s+by.+\(',
    'xwork.MethodAccessor',
'(?:define|eval|file_get_contents|include|require|require_once|shell_exec|phpinfo|system|passthru|preg_\w+|execute|echo|print|print_r|var_dump|(fp)open|concat|alert|showmodaldialog)\(',
    'xwork\.MethodAccessor',
    '(gopher|doc|php|glob|file|phar|zlib|ftp|ldap|dict|ogg|data)\:\/',
    'java\.lang',
    '\$_(GET|post|cookie|files|session|env|phplib|GLOBALS|SERVER)\[',
    '\<(iframe|script|body|img|layer|div|meta|style|base|object|input)',
    '(onmouseover|onerror|onload)\=',
    '\.(bak|inc|old|mdb|sql|backup|java|class)$',
    '\.(svn|htaccess|bash_history)',
    '(vhost|bbs|host|wwwroot|www|site|root|hytop|flashfxp).*\.rar',
    '(phpmyadmin|jmx-console|jmxinvokerservlet)',
    '/xmlrpc.php',
    '/\.git/config',
'/(attachments|upimg|images|css|uploadfiles|html|uploads|templets|static|template|data|inc|forumdata|upload|includes|cache|avatar)/(\w+).(php|jsp|asp)',

);

sub tail {

    my ( $filename, $linenum ) = @_;
    print "DEBUG :: tail() :: IN : $filename,$linenum \n" if $DEBUG;
    my $bw = File::ReadBackwards->new($filename)
      or die "can't read $filename $!";
    $linenum=1000 unless $linenum;
    my $count = 0;
    my @lines;

    while ( defined( my $line = $bw->readline ) ) {
        push @lines, $line;
        $count++;
        if ( $count == $linenum ) { last }
    }

    @lines = reverse @lines;
    return \@lines;
}

sub initCount {

    my $line   = shift;
    my @re     = @validurl;
    my $kcount = shift;
    my ( $zcount, $zip, $zrequrl, $zstatus, $siteurl );
    my $rawlog;

    for (@re) {
        my $result = scarlog1( $_, $line );
        my ( $mycount, $mylog ) = count($result);
        my $key = $_;
        $rawlog .= $mylog->{$key} if $mylog->{$key};

        $zcount += $mycount->{$key}->[0] if $mycount->{$key}->[0];
        print
          "DEBUG\:: initCount()\::OUT  $key $mycount->{$key}->[0]   $zcount \n"
          if $DEBUG;
        $zip->{$_} += $mycount->{$key}->[1]->{$_}
          for ( keys %{ $mycount->{$key}->[1] } );
        $zrequrl->{$_} += $mycount->{$key}->[2]->{$_}
          for ( keys %{ $mycount->{$key}->[2] } );

        if ($DEBUG) {
            print
"DEBUG\:: initCount()\::OUT  $key $zrequrl->{$_}  $_\=> $mycount->{$key}->[2]->{$_} \n"
              for ( keys %{ $mycount->{$key}->[2] } );
        }
        $zstatus->{$_} += $mycount->{$key}->[3]->{$_}
          for ( keys %{ $mycount->{$key}->[3] } );
        $siteurl->{$_} += $mycount->{$key}->[4]->{$_}
          for ( keys %{ $mycount->{$key}->[4] } );

    }
    if ($DEBUG) {
        print "DEBUG\:: initCount()\::OUT\::\$zrequrl  $_\=>$zrequrl->{$_}\n"
          for ( keys %{$zrequrl} );
    }
    return ( $rawlog, $zcount, $zip, $zrequrl, $zstatus, $siteurl );
}

sub count {

    my $result = shift;

    my ( $mcount, %rawlog );
    my $count = 0;
    for ( keys %{$result} ) {
        my ( %ip, %requrl, %status, %siteurl );

        next if $result->{$_} eq "";
        $rawlog{$_} .= $result->{$_};
        my @seclogs = split /\n/ms, $result->{$_};
        for (@seclogs) {
            $count++;
            print "DEBUG\:: count()\::IN $_\n" if $DEBUG;
            my ( $ip, $requrl, $status, $siteurl ) = (split)[ 0, 6, 8, 10 ];
            $ip{$ip}++           if $ip;
            $requrl{$requrl}++   if $requrl;
            $status{$status}++   if $status;
            $siteurl{$siteurl}++ if $siteurl;
            print
"DEBUG\:: count()\::OUT $ip\=>$ip{$ip} $requrl\=>$requrl{$requrl} $status\=>$status{$status} $siteurl\=>$siteurl{$siteurl} \n"
              if $DEBUG;
        }

        $mcount->{$_} = [ $count, \%ip, \%requrl, \%status, \%siteurl ];

    }

    return $mcount, \%rawlog;
}

sub scarlog1 {

    my ( $patter, $lines ) = @_;

    my %result;

    my $code = 'for(@{$lines}) {';
    $code .= 'if (m#';
    $code .= qr($patter);
    $code .= '#) {$result{' . q($patter) . '}.=$_}}';
    eval $code;
    die "Error ---: $@\n Code:\n$code\n" if ($@);

    #print "DEBUG scarlog1 :: OUT :: $_: $result{$_}\n" for(keys %result);
    return \%result;
}

sub iptabBan {

# must be root user;
# 必须root用户才可以操作iptables，当然也必须有iptables服务跑动着

    my $IP = shift;

    my $ips     = `/sbin/iptables-save`;
    my @ipsline = split /\n/sm, $ips;
    my $dist    = 0;
    for (@ipsline) {

        $dist = 1 if ( /$IP/ and /INPUT/ and /DROP/ );

    }
    unless ($dist) {
        `/sbin/iptables -I INPUT -s $IP -j DROP`;
        my $btime = localtime( time() );
        print "$btime :band $IP \n";
    }
    else {

        print "band alread!\n";

    }

}


sub nginxBan {

    my $btime = localtime( time() );
    my ( $ip, $conf, $pid ) = @_;
    my $bid = 0;
    open my $nFD, "<", $conf or die("Can not open the file!$!\n");
    while (<$nFD>) {
        print "DEBUG ::nginxBan :: $conf IN $_" if $DEBUG;
        $bid = 1 if /$ip/;
    }
    close $nFD;


    unless ($bid) {
        print "$btime,banip $ip\n";
        print $nFD "deny $ip\;\n";
        $pid = `cat $pid`;
        chomp $pid;
        `/usr/bin/kill -HUP $pid`;
    }


}

=head1 AUTHOR

ORANGE, C<< <bollwarm at ijz.me> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-app-waf at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-Waf>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::Waf


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-Waf>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/App-Waf>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/App-Waf>

=item * Search CPAN

L<http://search.cpan.org/dist/App-Waf/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 ORANGE.

This is free software; you can redistribute it and/or modify
it under the same terms as the Perl 5 programming language system itself.

=cut

1;    # End of App::Waf
